### PR TITLE
Implement object pooling to avoid memory leaks

### PR DESCRIPTION
**Pull request by ProGM**
Taking a look to the Unity Profiler, I noticed some memory leaks when using these outline effects.

I implemented object pooling, using the same approach Unity did in its UI

https://bitbucket.org/Unity-Technologies/ui

Hope it helps ;)

### CODE REVIEW

File: Assets/VertexEffects/Scripts/ListPool.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+static class ListPool<T>
+{
+        // Object pool to avoid allocations.
+ 	private static readonly ObjectPool<List<T>> s_ListPool = new ObjectPool<List<T>>(null, l => l.Clear());
```
Comment (n-yoda) [inline Assets/VertexEffects/Scripts/ListPool.cs]: Please use spaces for indentation.

File: Assets/VertexEffects/Scripts/ModifiedShadow.cs [hunk 2]
```diff
@@ -59,7 +59,7 @@ public void ModifyMesh(VertexHelper vh)
         if (!this.IsActive())
             return;
         
-        List<UIVertex> list = new List<UIVertex>();
+		var list = ListPool<UIVertex>.Get();
```
Comment (n-yoda) [inline Assets/VertexEffects/Scripts/ModifiedShadow.cs]: tabs -> spaces

File: Assets/VertexEffects/Scripts/ModifiedShadow.cs [hunk 3]
```diff
@@ -68,6 +68,7 @@ public void ModifyMesh(VertexHelper vh)
         vh.Clear();
 #endif
         vh.AddUIVertexTriangleStream(list);
+		ListPool<UIVertex>.Release(list);
```
Comment (n-yoda) [inline Assets/VertexEffects/Scripts/ModifiedShadow.cs]: tabs -> spaces

**Comment by n-yoda** [comment]
It seems you replaced `ModifiedShadow.ApplyShadow` with `Shadow.ApplyShadowZeroAlloc`.
It's correct if we use Unity >= 5.4, because Unity fixed the bug of capacity calculation:
https://bitbucket.org/Unity-Technologies/ui/diff/UnityEngine.UI/UI/Core/VertexModifiers/Shadow.cs?diff2=4bccdca73cf7&at=5.4

However, for Unity < 5.4, ModifiedShadow.ApplyShadow is better.
So we should implement a workaround using `#if UNITY_5_1 || UNITY_5_2 || UNITY_5_3`.

I will be glad if you will do this workaround, but it's ok for me to implement it after merging this PR.

**Comment by ProGM** [comment]
@n-yoda I did it! Please take a look ;)

(sorry for indents, I misconfigured Mono Develop ._.)

**Comment by n-yoda** [comment]
@ProGM Thank you!

**Comment by n-yoda** [comment]
After merging, I found UNITY_5_4_OR_NEWER directive. So I rewrote the workaround for versions in a simpler manner.

**Comment by ProGM** [comment]
👍

**Pull request merged**